scalafmt-dynamic/src/main/scala/org/scalafmt/dynamic/utils/ReentrantCache.scala

```diff
@@ -0,0 +1,61 @@
+package org.scalafmt.dynamic.utils
+
+import scala.annotation.tailrec
+import scala.concurrent.Await
+import scala.concurrent.Future
+import scala.concurrent.Promise
+import scala.concurrent.duration._
+import scala.util.Try
+
+class ReentrantCache[K, V] {
+  private[this] var cache: Map[K, Future[V]] = Map.empty
+
+  @tailrec
+  final def getOrAddToCache(key: K, shouldEvict: V => Boolean = _ => false)(get: () => V): V =
+    synchronized { // try to exit quickly from synchronized block
+      cache.get(key) match {
+        case Some(fut) => Right(fut)
+        case None =>
+          val p = Promise[V]()
+          cache += key -> p.future
+          Left(p)
+      }
+    } match {
+      case Right(fut) =>
+        /*fut.value match {
```

Comment:
I left the printlns in for now but can remove them if required.

Reply:
I normally don't leave printlns in the code but I don't have a strong preference against them. Let's remove it here for consistency with the rest of the codebase